tests/NATS.Client.Platform.Windows.Tests/TlsTests.cs

```diff
@@ -0,0 +1,150 @@
+    public async Task Tls_with_certificates()
+    {
+        const string caCertPem = "resources/certs/ca-cert.pem";
+        const string clientCertPem = "resources/certs/client-cert.pem";
+        const string clientKeyPem = "resources/certs/client-key.pem";
+
+        var tlsOpts = new NatsTlsOpts
+        {
+            RemoteCertificateValidationCallback = (_, certificate, _, errors) => ValidateCert(errors, certificate, LoadCaCerts(caCertPem)),
+            LoadClientCerts = () =>
+            {
+                var collection = LoadClientCerts(clientCertPem, clientKeyPem);
+                return new ValueTask<X509Certificate2Collection>(collection);
+            },
+        };
+
+        await using var server = await NatsServerProcess.StartAsync(config: "resources/configs/tls.conf");
+        await using var nats = new NatsConnection(new NatsOpts { Url = server.Url, TlsOpts = tlsOpts });
+
+        await nats.PingAsync();
+
+        await using var sub = await nats.SubscribeCoreAsync<int>("foo");
+        for (var i = 0; i < 64; i++)
+        {
+            await nats.PublishAsync("foo", i);
+            Assert.Equal(i, (await sub.Msgs.ReadAsync()).Data);
+        }
+    }
+
+    private static bool ValidateCert(SslPolicyErrors errors, X509Certificate certificate, List<X509Certificate2> trustedCaCerts)
+    {
+        if (errors == SslPolicyErrors.None)
+            return true;
+
+        var cert2 = new X509Certificate2(certificate);
+
+        foreach (var caCert in trustedCaCerts)
+        {
+            var chain = new X509Chain();
+            chain.ChainPolicy.ExtraStore.Add(caCert);
+            chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+            chain.ChainPolicy.VerificationFlags = X509VerificationFlags.AllowUnknownCertificateAuthority;
+
+            if (!chain.Build(cert2))
+                continue;
+            var root = chain.ChainElements[^1].Certificate;
+            if (root.Thumbprint == caCert.Thumbprint)
+                return true;
+        }
+
+        return false;
+    }
+
+    private static List<X509Certificate2> LoadCaCerts(string pem)
+    {
+        List<X509Certificate2> caCertificates = [];
+
+        using var reader = File.OpenText(pem);
+        var pemReader = new PemReader(reader);
+        while (pemReader.ReadObject() is { } obj)
+        {
+            if (obj is not Org.BouncyCastle.X509.X509Certificate bcCert)
+                continue;
+            var cert = new X509Certificate2(bcCert.GetEncoded());
+            caCertificates.Add(cert);
+        }
+
+        return caCertificates;
+    }
+
+    private static X509Certificate2Collection LoadClientCerts(string certPemPath, string keyPemPath)
+    {
+        var collection = new X509Certificate2Collection();
+
+        var certPem = File.ReadAllText(certPemPath);
+        var certParser = new X509CertificateParser();
+        var cert = certParser.ReadCertificate(new MemoryStream(Encoding.ASCII.GetBytes(certPem)));
```

Comment:
... On one hand, I would -think- that `ReadAllBytes` should work here as PEM files themselves -should- be ASCII (or perhaps 'UTF8 but in ASCII space so who notices'.

OTOH, it's a test so 🤷‍♂️ 

And TBH this approach looks pretty slick overall, I need to double-check but I _think_ it's way less 'hazardous' than loading the PEM directly (if you do that, at least on windows the risk of getting random zombie cert bytes in a local store go way up if the lifetime isn't properly managed... [interesting read](https://snede.net/the-most-dangerous-constructor-in-net/).)

Might need to borrow some of this elsewhere regardless,  😉 .

Reply:
done